pkg/azdiskschedulerextender/azdiskschedulerextender.go

```diff
@@ -0,0 +1,235 @@
+		}
+
+		// for every volume the pod needs, append its azVolumeAttachment name to the node name
+		for _, attachedVolume := range azVolumeAttachmentsMeta.volumes.Items {
+			_, needs := volumesPodNeeds[attachedVolume.Spec.AzVolumeName]
+			if needs {
+				nodeNameToVolumeMap[attachedVolume.Spec.AzDriverNodeName] = append(nodeNameToVolumeMap[attachedVolume.Spec.AzDriverNodeName], attachedVolume.Name)
+			}
+		}
+
+		// score nodes based in how many azVolumeAttchments are appended to its name
+		klog.V(2).Infof("Scoring nodes for pod %+v.", schedulerExtenderArgs.Pod)
+		for _, node := range availableNodes {
+			score := getNodeScore(len(nodeNameToVolumeMap[node.Name]), nodeNameToHeartbeatMap[node.Name])
+			hostPriority := schedulerapi.HostPriority{Host: node.Name, Score: score}
+			priorityList = append(priorityList, hostPriority)
+			klog.V(2).Infof("Score for %+v is %d.", node, score)
+		}
+	}
+	return
+}
+
+func getKubernetesClient() (clientSet.AzVolumeAttachmentInterface, clientSet.AzDriverNodeInterface, error) {
+
+	config, err := rest.InClusterConfig()
+	if err != nil {
+		// fallback to kubeconfig
+		kubeConfigPath := os.Getenv("KUBECONFIG")
+		if len(kubeConfigPath) == 0 {
+			kubeConfigPath = filepath.Join(os.Getenv("HOME"), ".kube", "config")
+		}
+
+		// create the config from the path
+		config, err = clientcmd.BuildConfigFromFlags("", kubeConfigPath)
+		if err != nil {
+			return nil, nil, fmt.Errorf("Cannot load kubeconfig: %v", err)
+		}
+	}
+
+	// generate the clientset extension based off of the config
+	azKubeExtensionClient, err := clientSet.NewForConfig(config)
+	if err != nil {
+		return nil, nil, fmt.Errorf("Cannot create clientset: %v", err)
+	}
+	azVolumeAttachmentExtensionClient := azKubeExtensionClient.AzVolumeAttachments("")
+	azDriverNodeExtensionClient := azKubeExtensionClient.AzDriverNodes("")
+
+	klog.Info("Successfully constructed kubernetes client and extension clientset")
+	return azVolumeAttachmentExtensionClient, azDriverNodeExtensionClient, nil
+}
+
+func getNodeScore(volumeAttachments int, heartbeat string) int64 {
+	// TODO: prioritize disks with low additional volume attachments
+	now := time.Now()
+	latestHeartbeatWas, err := time.Parse(time.UnixDate, heartbeat)
+	if err != nil {
+		return 0
+	}
+
+	latestHeartbeatCanBe := now.Add(-2 * time.Minute)
+	klog.V(2).Infof("Latest node heartbeat was: %s. Latest accepted heartbeat can be: %s", heartbeat, latestHeartbeatCanBe.Format(time.UnixDate))
+
+	if latestHeartbeatWas.Before(latestHeartbeatCanBe) {
+		return 0
+	}
+
+	score := int64(volumeAttachments*100) - int64(now.Sub(latestHeartbeatWas)/10000000000) //TODO fix logic when all variables are finalized
+	return score
+}
+
+func getAzDriverNodes(context context.Context, out chan azDriverNodesMeta) {
+	// get all nodes that have azDriverNode running
+	var activeDriverNodes azDriverNodesMeta
+	activeDriverNodes.nodes, activeDriverNodes.err = azDriverNodeExtensionClient.List(context, metav1.ListOptions{})
```

Comment:
Not sure if this will return all the labels. We should return all the labels, that will have zone\fd information.